File: 003/python/003.py
# Problem 3
# Largest prime factor

import sys


def main():
    target = 600851475143

    if len(sys.argv) > 1:
        target = int(sys.argv[1])

    print(factor(target))


def factor(n):
    fac = n

    for i in range(2, int(n / 2)):
        if n % i == 0:
            fac = i
            break

    if fac == n:
        return fac

    new = factor(int(n / fac))
    return fac if fac > new else new


main()

File: 003/c/003.c
/* Problem 3
   Largest prime factor
*/

#include <stdio.h>
#include <stdlib.h>

long factor(long n);

int main(int argc, char *argv[]) {
    long target = 600851475143;
    char *ptr;

    if (argc > 1) target = strtol(argv[1], &ptr, 10);

    printf("%li\n", factor(target));
}

long factor(long n) {
    long fac = n;

    for (long i = 2; i < n / 2; i++) {
        if (n % i == 0) {
            fac = i;
            break;
        }
    }

    if (fac == n) return fac;

    long new = factor(n / fac);
    return fac > new ? fac : new;
}
